### DISCUSSION TITLE
Inconsistent spacing with repeated boxes in outline

### DISCUSSION THREAD

**Discussion by Victor-N-Suadicani** [typst/typst#4160] [Q&A] [answered]
I'm trying to make a pretty outline like this:

```typst
#set heading(numbering: "1.")

#show outline.entry: it => {
    // First level is bold and has spacing.
    if it.level == 1 {
        v(1.5em, weak: true)
    }

    let entry = {
        link(
            it.element.location(),
            {   
                it.body.at("children").at(0)
                h(0.5cm)
                it.body.at("children").slice(2).join()
            }
        )

        h(0.25cm)
        box(width: 1fr, it.fill)
        link(it.element.location(), box(width: 0.5cm, align(end, it.page)))
    }

    if it.level == 1 {
        set text(size: 14pt)
        strong(entry)
    } else {
        entry
    }
}

Annoying dots not lining up: $arrow.b$

#outline(depth: 3, indent: true, fill: repeat(box(width: 1cm / 3, fill: rgb(0, 0, 0, 32), ".")))

= Heading 1
#lorem(1000)
= Heading 2
#lorem(1000)
== Heading 3
#lorem(1000)
== Heading 4
#lorem(1000)
= Heading 5
#lorem(1000)
== Heading 6
#lorem(1000)
== Heading 7
#lorem(1000)
=== Heading 8
#lorem(1000)
== Heading 9
#lorem(1000)
=== Heading 10
#lorem(1000)
```

This results in the following:
![image](https://github.com/typst/typst/assets/90071638/0ccc3e1c-3237-4723-b7bc-86a557ff2609)

And it's mostly good _except_ that the dots somehow don't line up! They always start lined up on the right (which looks good) but then as you go to the left of the page, they somehow get spaced out inconsistently.

If I give the boxes a background color, I can even see that there is sometimes a _tiny_ gap in-between the boxes.

![image](https://github.com/typst/typst/assets/90071638/5f422f43-a1c5-4a0b-bbe0-0e23985c2d70)

My guess is that this gap is what is causing the problem but I'm not sure.

Am I causing this spacing somehow and can I fix it? Or is this some kind of bug?

**Comment by l0wx** [comment, marked as answer]
```typst
#import calc: *

#let demo(sz) = place(rect(..sz, stroke: green))// Demonstrational purp.
#let filler = box(
    fill: red,// Demonstrational purp.
    ".",
)
#let filler-seq = context {
    let msms = measure(filler)

    box(
        width: 1fr,
        inset: (left: -msms.width/2),
        align(
            right,
            layout(sz=> demo(sz) + context filler*(msms, sz)
                .map(l=> 1000000 * l.width.pt())
                .fold(1, (acc, r)=> quo(r, acc))
            ),
        ),
    )
}

#outline(
    fill: filler-seq,
)

= Ans
= Dswa
```

**Reply by Victor-N-Suadicani** [reply]
Thanks, this works better!

Could you explain this solution? I'm not too familiar with Typst to know really what's going on here.

Also, do you have any idea what went wrong with my original version?

**Reply by l0wx** [reply]
1. It contextually obtains the available space via `layout()` and repeats the filler accordingly based on a division of the available width by the `measure()`red (`msms` variable) width of the filler. `inset: (left: -msms.width/2)` is just a refinement to prevent the spacing between the heading text and the fillers from getting noticeably big by allowing for another repetition.
2. If the remain doesn't allow for yet another repetition, everything is widened to fill the whole available space:
> Space may be inserted between the instances of the body parameter, so be sure to include negative space if you need the instances to overlap.
—https://typst.app/docs/reference/layout/repeat/
